### PR TITLE
GRW-1914 / Feature / Prevent duplicate creation of shop session

### PR DESCRIPTION
**Pull request by alebedev**
## Describe your changes

Refactor ShopSessionProvider to work effectively inall environments: SSR, SSG, client

## Justify why they are needed

Prevent duplicate session creation and session requests

## Checklist before requesting a review

- [ ] I have performed a self-review of my code

### CODE REVIEW

**Review by robinandeer — APPROVED** [review]
Strong persistent work here 🦾

Not 100% sure it's easier to reason about this code but willing to give it a go! Nice to see that we can move some things outside React but not thrilled about adding `useEffect` to kick off data fetching 😅

File: apps/store/src/services/shopSession/ShopSessionContext.tsx [hunk 1]
```diff
@@ -1,61 +1,25 @@
...
-  // Has to be wrapped to prevent duplicate execution (Apollo quirk leads do duplicate execution when called directly from render)
+  const shopSessionId = contextValue?.data?.shopSession.id
   useEffect(() => {
-    // TODO: isBrowser() and ensure code splitting does not break
-    if (typeof window !== 'undefined' && !shopSessionId && !mutationResult.called) {
-      createShopSession()
-    }
-  }, [createShopSession, mutationResult.called, shopSessionId])
+    Track.addContext('shopSessionId', shopSessionId)
+  }, [shopSessionId])
```
[the remark this answers is not in the document]
Comment (alebedev) [reply apps/store/src/services/shopSession/ShopSessionContext.tsx]: We need this on client, session creation _may_ happen on server.  Hook looks like lesser of evils here.  Removed it from the provider and made explicit

File: apps/store/src/services/shopSession/ShopSessionService.ts [hunk 2]
```diff
@@ -8,38 +8,48 @@ import {
...
+  public async getOrCreate(params: ShopSessionCreateMutationVariables): Promise<ShopSession> {
     const existingShopSession = await this.fetch(params.locale as IsoLocale)
 
     if (existingShopSession?.countryCode === params.countryCode) {
       return existingShopSession
     }
 
-    return await this.create(params)
+    // Deduplicate mutation, Apollo won't do this for us
+    if (!this.createPromise || JSON.stringify(this.createParams) != JSON.stringify(params)) {
```
[the remark this answers is not in the document]
Comment (alebedev) [reply apps/store/src/services/shopSession/ShopSessionService.ts:45]: Typescript warns about weak comparisons when it detects them.  Here it knows that left and right side has the same type, therefor `!=` is safe

File: apps/store/src/services/shopSession/ShopSessionContext.tsx [hunk 3]
```diff
@@ -79,7 +43,35 @@ export const useShopSession = () => {
...
+  const { countryCode } = useCurrentCountry()
+  const { locale } = useCurrentLocale()
   const apolloClient = useApolloClient()
-  return setupShopSessionServiceClientSide(apolloClient)
+  // Only used client-side
+  const shopSessionService = useMemo(
+    () => setupShopSessionServiceClientSide(apolloClient),
+    [apolloClient],
+  )
+  const shopSessionId = shopSessionService.shopSessionId() ?? initialShopSessionId
```
[the remark this answers is not in the document]
Comment (alebedev) [reply apps/store/src/services/shopSession/ShopSessionContext.tsx:48]: No, in that case you'll be stuck with old session from first render even if new one was already created

File: apps/store/src/services/shopSession/ShopSessionContext.tsx [hunk 4]
```diff
@@ -79,7 +43,35 @@ export const useShopSession = () => {
...
+
+  const queryResult = useShopSessionQuery({
+    variables: shopSessionId ? { shopSessionId, locale } : undefined,
+    skip: !shopSessionId,
+  })
+
+  useEffect(() => {
+    if (isBrowser()) {
+      if (initialShopSessionId && shopSessionService.shopSessionId() !== initialShopSessionId) {
+        shopSessionService.saveId(initialShopSessionId)
```
[the remark this answers is not in the document]
Comment (alebedev) [reply apps/store/src/services/shopSession/ShopSessionContext.tsx:58]: Will break if created on server

[the remark this answers is not in the document]
Comment (alebedev) [reply apps/store/src/services/shopSession/ShopSessionContext.tsx:58]: I'll check if we can rely on cookies and drop this completely

File: apps/store/src/pages/checkout/index.tsx [hunk 5]
```diff
@@ -15,17 +15,18 @@ import {
...
 type NextPageProps = Omit<CheckoutPageProps, 'loading' | 'userError'> & {
   checkoutId: string
   checkoutSigningId: string | null
-  shopSessionId: string
 }
 
 const NextCheckoutPage: NextPage<NextPageProps> = (props) => {
-  const { products, checkoutId, checkoutSigningId, shopSessionId, ...pageProps } = props
+  const { products, checkoutId, checkoutSigningId, ...pageProps } = props
+  const { shopSession } = useShopSession()
```
[the remark this answers is not in the document]
Comment (alebedev) [reply apps/store/src/pages/checkout/index.tsx:29]: This is already handled in `getServerSideProps`, page component may use generic hook without bothering with the source of shopSession data

File: apps/store/src/services/shopSession/ShopSessionService.ts [hunk 6]
```diff
@@ -8,38 +8,48 @@ import {
...
+  private createParams: ShopSessionCreateMutationVariables | null = null
+  private createPromise: Promise<ShopSession> | null = null
+
   public shopSessionId() {
     return this.persister.fetch()
   }
 
-  public save(shopSession: ShopSession) {
-    this.persister.save(shopSession.id)
+  public saveId(shopSessionId: string) {
```
[the remark this answers is not in the document]
Comment (alebedev) [reply apps/store/src/services/shopSession/ShopSessionService.ts:27]: On the other hand, when you see `save(shopSession)` you expect full object to be persisted

**Pull request merged**